### PR TITLE
feat: [PL-39005]: bijou implementation delegate side

### PR DESCRIPTION
**Pull request by xingchi-jin**
## Describe your changes
This change does
1. Acquire task payload using the new API created at PR https://github.com/harness/harness-core/pull/48331 
2. based on different type of request init or exec or clean, it invokes different handlers 
3. Based on runner type, it invokes different runners. 
4. Added a runner factory so we can easily add more runners. defined a common runner interface, which essentially takes binary input

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49272)

### CODE REVIEW

**Review by mdmilic — COMMENTED** [review]
> Reviewed 2 of 22 files at r1, 48 of 48 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49272)* status: Some of the checks are too old. 50 file(s) to review. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`WORKSPACE` line 4914 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYYo8s92Sxt9shaD3Oi:-NYYo8sA86Cijyg7B0Dz:b-i4yt4r) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/WORKSPACE#L4914)):*
> ```Starlark
>     "guru.nidi:graphviz-java:0.16.3",
>     "guru.nidi:graphviz-rough:0.16.3",
>     "tv.cntt:jauter:1.7",
> ```

I don't think you should be able to add dep like this. Any WORKSPACE changes need to go through APPSEC.
I think this is the current process https://harness.atlassian.net/wiki/spaces/PD/pages/1932230714/3rd+Party+Libraries#Review-Request-Process:
And here's a recent example: https://harness.atlassian.net/browse/SECAPP-367

re-evaluate if you really need it. Seems unmaintained (latest version from 2014). 
___
*[`260-delegate/src/main/java/io/harness/delegate/app/DelegateAgentApplication.java` line 169 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYYpRKC0I6l6oEdEZpn:-NYYpRKC0I6l6oEdEZpo:b-zg3hx9) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/app/DelegateAgentApplication.java#L169)):*
> ```Java
>     environment.jersey().register(injector.getInstance(ExecutionResponseResource.class));
>     environment.jersey().register(injector.getInstance(DelegateAgentMetricResource.class));
>     environment.jersey().register(injector.getInstance(KryoFeature.class));
> ```

Why KryoFeature?
___
*[`260-delegate/src/main/java/io/harness/delegate/app/ExecutionResponseResource.java` line 40 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYYpcIgDH48TVUBdMjg:-NYYpcIgDH48TVUBdMjh:b-lzgav) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/app/ExecutionResponseResource.java#L40)):*
> ```Java
> 
>   @POST
>   @Consumes("application/x-kryo-v2")
> ```

Should not be Kryo, should be proto. This is an existing proto wrapper intended for sending the response. Try using that and modify if needed. https://github.com/harness/harness-core/blob/develop/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/plugin_execution_status.proto

Also, some version of this class already exists, see if you can reuse or combine into a single one
https://github.com/harness/harness-core/blob/develop/260-delegate/src/main/java/io/harness/delegate/service/core/resource/TaskResource.java
___
*[`260-delegate/src/main/java/io/harness/delegate/app/ExecutionResponseResource.java` line 50 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYYqcZODC39N46ao9yy:-NYYqcZODC39N46ao9yz:b-2rk5kv) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/app/ExecutionResponseResource.java#L50)):*
> ```Java
>     } catch (IOException e) {
>       // TODO: define exception (?? extends WingsException)
>       e.printStackTrace();
> ```

use @Slf4j
___
*[`260-delegate/src/main/java/io/harness/delegate/app/modules/platform/k8s/K8SRunnerModule.java` line 25 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYYrddM3D5biNrnuO2p:-NYYrddM3D5biNrnuO2q:b-ezjmg6) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/app/modules/platform/k8s/K8SRunnerModule.java#L25)):*
> ```Java
>      */
>     install(new ApiClientModule());
>     bind(Runner.class).annotatedWith(Names.named("k8sRunner")).to(K8SLiteRunner.class);
> ```

There wouldn't be different runner type instances baked into the Immutable Delegate
___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 958 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_B9iYEiyelMIw-g_D:-NY_B9iYEiyelMIw-g_E:b89wsge) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L958)):*
> ```Java
>       try {
>         if (!acquireTasks.get()) {
>           log.info("[Old] Upgraded process is running. Won't acquire task while completing other tasks");
> ```

This is not needed, it's for legacy watcher upgrade scenario. I think we should only support this new architecture with immutable delegates. Or at least the log is incorrect (if there are other use cases to stop acquire - e.g. during shutdown)
___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 970 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_Bh3aDurfVJPduVNj:-NY_Bh3aDurfVJPduVNk:b2ni78f) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L970)):*
> ```Java
>         handlerMappingServer.serve(acquired);
>       } catch (IOException e) {
>         e.printStackTrace();
> ```

log
___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 977 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_8qAu1rmK9I2Gx1Ou:-NY_8qAu1rmK9I2Gx1Ov:b2ytbq6) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L977)):*
> ```Java
>   private boolean shouldProceedProcessingTaskEvent(String taskId) {
>     if (taskId == null) {
>       log.warn("Delegate state machine id cannot be null");
> ```

no state machine
___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 2026 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_8c-97s4Sn5RvHcIF:-NY_8c-97s4Sn5RvHcIG:b-4bygje) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L2026)):*
> ```Java
>     //    }
>     //
>     //    if (rejectRequest.get()) {
> ```

remove commented code
___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 2027 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_9k4UBFr2DuFjS7fm:-NY_9k4UBFr2DuFjS7fn:b-koqtrm) ([raw file](https://github.com/harness/harness-core/blob/0e8e26b6dfb2e16c708f7c252c77634045ff5aa6/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L2027)):*
> ```Java
>         }
> 
>         if (currentlyAcquiringTasks.contains(delegateTaskId)) {
> ```

I think this one is needed, looks like you just removed it
___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 2090 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_A6VbAK-8q0VrpDFl:-NY_A6VbAK-8q0VrpDFm:b-gaa3jr) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L2090)):*
> ```Java
> 
>         // Delegate assignment protocol implicitly use instance id (set or not) to determine if a task has been
>         // assigned. This is not a good way of design. Good way is to use delegate id, or precisely one of delegate id
> ```

instanceId doesn't really make sense, it's a legacy thing. We should just look to completely get rid of it.
___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 2694 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_AGv_DavpfYGxVlbS:-NY_AGv_DavpfYGxVlbT:b-z81cld) ([raw file](https://github.com/harness/harness-core/blob/0e8e26b6dfb2e16c708f7c252c77634045ff5aa6/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L2694)):*
> ```Java
>         char[] secretValue = decryptedRecords.get(value.getEncryptedRecord().getUuid());
>         if (secretValue == null) {
>           throw new UnexpectedException(format("Value for secret [%s] (uuid: [%s]) found null.",
> ```

Why removing this?
___
*[`260-delegate/src/main/java/io/harness/delegate/service/common/AcquireTaskHelper.java` line 26 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_CeBC127JgH03OAgR:-NY_CeBC127JgH03OAgS:bvtnxi) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/common/AcquireTaskHelper.java#L26)):*
> ```Java
> @Slf4j
> @Singleton
> public class AcquireTaskHelper {
> ```

The two acquire methods seem completely identical, why not just have one and pass the `Call<?>`
___
*[`260-delegate/src/main/java/io/harness/delegate/service/common/AcquireTaskHelper.java` line 28 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_Bo3i2jHTXc0V43Zo:-NY_Bo3i2jHTXc0V43Zp:b-feus3p) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/common/AcquireTaskHelper.java#L28)):*
> ```Java
> public class AcquireTaskHelper {
>     private final Set<String> currentlyAcquiringTasks = ConcurrentHashMap.newKeySet();
>     @Inject private DelegateAgentManagerClient managerClient;
> ```

no field injection & make `final`
___
*[`260-delegate/src/main/java/io/harness/delegate/service/common/AcquireTaskHelper.java` line 35 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_C7WsDIkJx_szyAED:-NY_C7WsDIkJx_szyAEE:b-ql1mqh) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/common/AcquireTaskHelper.java#L35)):*
> ```Java
>         final String delegateInstanceId,
>         final String taskId,
>         Consumer<Response<AcquireTasksResponse>> handler) throws IOException {
> ```

errorHandler
___
*[`260-delegate/src/main/java/io/harness/delegate/service/common/AcquireTaskHelper.java` line 38 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_CBPI8uQ8Y7jkrWbD:-NY_CBPI8uQ8Y7jkrWbE:b2ytbq6) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/common/AcquireTaskHelper.java#L38)):*
> ```Java
> 
>         if (currentlyAcquiringTasks.contains(taskId)) {
>           log.info("Task [Delegate state machine entity: {}] currently acquiring. Don't acquire again", taskId);
> ```

no state machine
___
*[`260-delegate/src/main/java/io/harness/delegate/service/handlermapping/HandlerMappingServer.java` line 28 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_6Ulz0CyCsFuY5hts:-NY_6Ulz0CyCsFuY5htt:b-puv7uf) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/handlermapping/HandlerMappingServer.java#L28)):*
> ```Java
> @Slf4j
> @Singleton
> public class HandlerMappingServer {
> ```

Honestly I think this router thing is completely unnecessary. It just introduces additional layer of abstraction and makes things more complex than they should be and is generally harder to understand. It seems you can achieve the same thing with a simple if then else, or even by using RunnerFactory directly. 
Also, I'm not quite sure I understand how it would work for other runner types that are not baked with delegate and that you have to hit with an API call. 
Lastly, sometime hitting a runner might not even be an option (e.g. exec runner), I think this runner would have to phone back to delegate instead to acquire payload (i.e. delegate might not be able to reach the runner, but runner can reach delegate). 
___
*[`260-delegate/src/main/java/io/harness/delegate/service/handlermapping/HandlerMappingServer.java` line 31 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYZzyqr81-uOHp2Gopq:-NYZzyqr81-uOHp2Gopr:b-i53hid) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/handlermapping/HandlerMappingServer.java#L31)):*
> ```Java
>   private ThreadPoolExecutor taskExecutor;
>   AcquireTaskHelper acquireTaskHelper;
>   HandlerMappingRouter router;
> ```

make all `private final`
___
*[`260-delegate/src/main/java/io/harness/delegate/service/handlermapping/handlers/ExecutionHandler.java` line 23 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_0qjB7lITu2lpj2No:-NY_0qjB7lITu2lpj2Np:b-rcoewg) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/handlermapping/handlers/ExecutionHandler.java#L23)):*
> ```Java
> @Singleton
> public class ExecutionHandler implements Handler {
>   @Inject private RunnersFactory runnersFactory;
> ```

Don't do field injection. Only constructor injection. Also, make the field `final`
___
*[`260-delegate/src/main/java/io/harness/delegate/service/handlermapping/handlers/ExecutionInfrastructureHandler.java` line 28 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_1T_y1Mok0uCTXCyi:-NY_1T_y1Mok0uCTXCyj:b-9j8uw0) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/handlermapping/handlers/ExecutionInfrastructureHandler.java#L28)):*
> ```Java
> @Slf4j
> public class ExecutionInfrastructureHandler implements Handler {
>   @Inject RunnersFactory runnersFactory;
> ```

No field injection & make all fields private final
___
*[`260-delegate/src/main/java/io/harness/delegate/service/runners/RunnersFactory.java` line 30 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_2QOe39YSoCLb0Oem:-NY_2QOe39YSoCLb0Oen:bw80ddt) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/runners/RunnersFactory.java#L30)):*
> ```Java
>     public RunnersFactory(@Named("k8sRunner") final Runner k8SLiteRunner) {
>         // add default runners
>         addRunner(RUNNER_TYPE_K8S, k8SLiteRunner);
> ```

How would this work for other runner types? Only K8SLiteRunner is going to be baked into the delegate, all other runner types are separate executables. 
___
*[`260-delegate/src/main/java/io/harness/delegate/service/runners/RunnersModule.java` line 19 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_32-3Eyk0ddsmw_mS:-NY_32-4Ed3JM3LzY34J:b8nawq1) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/runners/RunnersModule.java#L19)):*
> ```Java
>     protected void configure() {
>         install(new K8SRunnerModule());
>         bind(RunnersFactory.class).in(Singleton.class);
> ```

You don't need this when you used `@Singleton` annotation (you can use one or the other)
___
*[`400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java` line 167 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYZirJ2B92r4YV5nDVj:-NYZirJ2B92r4YV5nDVk:bs9reuh) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/400-rest/src/main/java/io/harness/grpc/DelegateServiceGrpcImpl.java#L167)):*
> ```Java
>               .uuid(taskId)
>               .runnerType(schedulingConfig.getRunnerType())
>               .infraId(executionInfraRef.orElse(""))
> ```

Why not just omit it if you don't have it. 
___
*[`420-delegate-agent/src/main/java/io/harness/managerclient/DelegateAgentManagerClient.java` line 272 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYZk3Qp2ZW-D5SuPylh:-NYZk3Qp2ZW-D5SuPyli:bf0iavt) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/420-delegate-agent/src/main/java/io/harness/managerclient/DelegateAgentManagerClient.java#L272)):*
> ```Java
> 
>   @Consumes({"application/x-protobuf"})
>   @GET("executions/payload/{executionId}")
> ```

`executions/{executionId}` is just fine, no need for /payload
___
*[`420-delegate-agent/src/main/java/io/harness/managerclient/DelegateAgentManagerClient.java` line 274 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYZjsor2o3DEaOmHR5T:-NYZjsos4g5E_9_Y2nEQ:b-j7hahj) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/420-delegate-agent/src/main/java/io/harness/managerclient/DelegateAgentManagerClient.java#L274)):*
> ```Java
>   @GET("executions/payload/{executionId}")
>   Call<AcquireTasksResponse> acquireTaskPayload(@Path("executionId") String uuid, @Query("delegateId") String delegateId,
>                                                 @Query("accountId") String accountId, @Query("delegateInstanceId") String delegateInstanceId);
> ```

Do we still need instance id?
___
*[`420-delegate-agent/src/main/java/io/harness/managerclient/DelegateAgentManagerClient.java` line 277 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYZkiQ_6aWvWKhAVduI:-NYZkiQ_6aWvWKhAVduJ:blmhgcf) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/420-delegate-agent/src/main/java/io/harness/managerclient/DelegateAgentManagerClient.java#L277)):*
> ```Java
> 
>   @Consumes({"application/x-protobuf"})
>   @POST("executions/response/{executionId}/executionInfra")
> ```

Do we need a separate API endpoint for infra/exec/cleanup responses?
Remove `response` word, also we use kebab case for APIs. 
You can use something like `"executions/infra/{executionId}"`
___
*[`870-cg-orchestration/src/main/java/io/harness/beans/DelegateTask.java` line 113 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_3Ivq4WE7fUpr5xj0:-NY_3Ivq4WE7fUpr5xj1:b-e33m4t) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/870-cg-orchestration/src/main/java/io/harness/beans/DelegateTask.java#L113)):*
> ```Java
>   private String requestUri;
>   private String requestMethod;
>   private String infraId;
> ```

Is infraId only going to become available after DelegateTask has been already written to the DB (i.e. so you would have to do the update to write it)? If so I don't think it belongs to this collection. 
___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/acquire_task.proto` line 45 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYZiMG78nosFm91jzGj:-NYZiMG78nosFm91jzGk:b-rxnutn) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/acquire_task.proto#L45)):*
> ```proto
>   InputData task_data = 7;   // binary data for task inputs.
>   InputData infra_data = 8;  // binary data for infra config
>   string executionInfraId = 9;  // Id for the execution infra to use
> ```

This should not be inside TaskPayload. It's id for the entire task group, not per task. 
___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/plugin.proto` line 41 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYZm1Vc52xQ7qgGEinL:-NYZm1Vc52xQ7qgGEinM:b76yaub) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/plugin.proto#L41)):*
> ```proto
>   ResponseCode responseCode = 1;
>   string errorMessage = 2;
>   InfraLocationInfo location = 3;
> ```

rename field to something like `infra` or any single word that makes sense after you rename the message
___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/plugin.proto` line 44 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYZlibY0JkerVLcg-OK:-NYZlibY0JkerVLcg-OL:bpyfiqa) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/plugin.proto#L44)):*
> ```proto
> }
> 
> message InfraLocationInfo {
> ```

Let's not use word `Location` doesn't really have any meaning in this context. `InfraInfo`, `InfraDetails` etc... is fine I think.

**Review by xingchi-jin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49272)* status: 52 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/runners/RunnersFactory.java` line 30 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NY_2QOe39YSoCLb0Oem:-NYtyw2J8RIg7YzYn3wP:b-g98iax) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/260-delegate/src/main/java/io/harness/delegate/service/runners/RunnersFactory.java#L30)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

How would this work for other runner types? Only K8SLiteRunner is going to be baked into the delegate, all other runner types are separate executables. 
</blockquote></details>

IF other runners are executables, here just bundle the client class which makes the call to the actual runner 
___
*[`955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/acquire_task.proto` line 45 at r2](https://reviewable.io/reviews/harness/harness-core/49272#-NYZiMG78nosFm91jzGj:-NYtzXFr5HSFkh_o-PsF:bwawhe) ([raw file](https://github.com/harness/harness-core/blob/1bc9fab8024c3975afe991bfda2b4dcb755acf9d/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/acquire_task.proto#L45)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

This should not be inside TaskPayload. It's id for the entire task group, not per task. 
</blockquote></details>

i think both are fine for current implementation
if in polling mode, we acquire a list of tasks. the tasks could use different infras. Adding it inside taskPayLoad makes sense in an interpretation that this task is going to use this infra.

File: project/main_maven_install.json [hunk 1]
```diff
@@ -35839,6 +35839,22 @@
                 "sha256": "df6905a047b05e23bc91f03ba57ac2f87c1ddf83e048aa0e5bd13169d5ebf0d9",
                 "url": "https://harness-artifactory.harness.io/artifactory/portal-maven/stax/stax/1.2.0/stax-1.2.0.jar"
             },
+            {
+                "coord": "tv.cntt:jauter:1.7",
```
Comment (gn-harness) [inline project/main_maven_install.json:35843]: This is a very old library is not maintained - https://github.com/sinetja/jauter/issues. Any future vulnerability will not be fixable

Please find another library

File: 260-delegate/src/main/java/io/harness/delegate/service/common/ManagerCallHelper.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.function.Consumer;
+import lombok.experimental.UtilityClass;
+import lombok.extern.slf4j.Slf4j;
+import okhttp3.internal.http2.StreamResetException;
+import retrofit2.Call;
+import retrofit2.Response;
+
+@Slf4j
+@UtilityClass
+public class ManagerCallHelper {
```
Comment (gn-harness) [inline 260-delegate/src/main/java/io/harness/delegate/service/common/ManagerCallHelper.java:28]: These methods seem to be copied over from DelegateAgentServiceImpl.java - can you please converge these two.

Also, it is sad that the existing class also did not have any tests :( Feel free to add a TODO for tests

File: project/main_maven_install.json (hunk 1 above)
Comment (gn-harness) [reply project/main_maven_install.json:35843]: I see this is used for "this.router = new HandlerMappingRouter()
                      .POST("/executionInfrastructure/:runnerType", ExecutionInfrastructureHandler.class)
                      .POST("/execution/:runnerType", ExecutionHandler.class)
                      .DELETE("/executionInfrastructure/:runnerType", ExecutionInfrastructureDeleteHandler.class);"

-- feel free to simplify this into a map and then refer directly

File: 260-delegate/src/main/java/io/harness/delegate/service/handlermapping/HandlerMappingServer.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    var routed =
+        router.route(SchedulingTaskEvent.Method.valueOf(taskPayload.getResourceMethod()), taskPayload.getResourceUri());
+    if (routed == null || routed.notFound()) {
+      log.error("404 error for {} {}", taskPayload.getResourceMethod(), taskPayload.getResourceUri());
+      return;
+    }
+    var handlerClass = routed.target();
+    var handler = injector.getInstance(handlerClass);
+
+    var handlerContext = context.deepCopy();
```
Comment (gn-harness) [inline 260-delegate/src/main/java/io/harness/delegate/service/handlermapping/HandlerMappingServer.java:59]: nit: 

seems like too many "vars" being used. I think vars are useful when you are doing operations like:

void printFirstAnimal(List<Animal> listOfAnima) {
 var animal = listOfAnimals.get(0);
 System.out.println(animal);
}

In the above function, type of "animal" is very clear - as a function reader I don't have to wonder. However, here I had to think/infer, what would be type of "routed", handlerClass, handler, etc.

@mdmilic Any best practices you follow on this?

**Review by gn-harness — COMMENTED** [review]
Looks good, just one blocking comment on the library used.

We are publishing some guidelines (https://docs.google.com/document/d/1lN2UNelnvpw-FVgOo1-Heh8T80Wb-JOAyiSFZPWikO0/edit#heading=h.3vp6pfu2bmdn) on how to pick libraries

File: 260-delegate/src/main/java/io/harness/delegate/service/handlermapping/HandlerMappingServer.java (hunk 3 above)
Comment (xingchi-jin) [reply 260-delegate/src/main/java/io/harness/delegate/service/handlermapping/HandlerMappingServer.java:49]: never hurt to be verbose. fixed

File: project/main_maven_install.json (hunk 1 above)
Comment (xingchi-jin) [reply project/main_maven_install.json]: removed the usage of this lib

**Comment by xingchi-jin** [comment]
trigger uts

**Comment by xingchi-jin** [comment]
trigger uts

**Comment by xingchi-jin** [comment]
trigger ut9

File: 260-delegate/src/main/java/io/harness/delegate/service/common/ManagerCallHelper.java (hunk 2 above)
Comment (xingchi-jin) [reply 260-delegate/src/main/java/io/harness/delegate/service/common/ManagerCallHelper.java:25]: The common code has been moved here. the TODO is to add a unit test.

**Review by gn-harness — DISMISSED** [review]

**Comment by xingchi-jin** [comment]
trigger uts

**Comment by xingchi-jin** [comment]
trigger uts

**Review by gn-harness — APPROVED** [review]

**Comment by xingchi-jin** [comment]
trigger ut0

**Comment by xingchi-jin** [comment]
trigger ut5

**Comment by xingchi-jin** [comment]
trigger uts

**Comment by xingchi-jin** [comment]
trigger ut5

**Comment by xingchi-jin** [comment]
trigger ut9

**Comment by xingchi-jin** [comment]
trigger ut5

**Pull request merged**